### PR TITLE
Proposal for escaping within lexed content

### PR DESCRIPTION
**Pull request by jneen**
We often receive requests to do things such as underline sections of code, or embed arbitrary html inside of highlighted content. This PR is a proposal to address such features, using a similar tactic as Pygments.

We provide a special `Escape` lexer, which is capable of taking user-defined escape delimiters (default `<!` and `!>`) and rendering their contents into the `Escape` token (also from Pygments). In the HTML formatter (and potentially other formatters), we can allow the `Escape` token to pass as a raw value.

## IMPORTANT

There is a **major security concern** with this feature: in some applications (such as Gitlab), an untrusted user will control all of the input text, the lexer choice, and the lexer options. We cannot allow such users to pass unescaped HTML into the output.

For this reason, escaping is disabled by default - the base `Formatter` class will filter all `Escape` tokens and replace them with `Error` tokens before passing the token stream to the `stream` method. In order to enable this feature, a user must explicitly call `Rouge::Formatter.enable_escape!`. It can also be enabled thread-locally with `Rouge::Formatter.with_escape { ... }` and explicitly disallowed with `Rouge::Formatter.disable_escape!`. I **think** this should have the correct secure defaults and extension opportunities for most applications, but I would appreciate additional security review.

cc @stanhu

### CODE REVIEW

**Comment by ashmaroli** [comment]
@jneen Is the overlap with changes in #1151 a conscious move or mistake?

If yes, please add a "Depends on #1151" somewhere in the OP so that it is clear that this should be merged only after #1151

**Comment by jneen** [comment]
Good point, we do depend on #1151. I will rebase once that is merged.

File: lib/rouge/lexers/escape.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        @to_start_regex ||= /(.*?)(#{Regexp.escape(@start)})/m
+      end
+
+      def to_end_regex
+        @to_end_regex ||= /(.*?)(#{Regexp.escape(@end)})/m
+      end
+
+      def stream_tokens(str, &b)
+        stream = StringScanner.new(str)
+        p :to_start_regex => to_start_regex
```
[the remark this answers is not in the document]
Comment (ashmaroli) [reply lib/rouge/lexers/escape.rb]: Most likely *developmental artifacts*.. :smile:

[the remark this answers is not in the document]
Comment (jneen) [reply lib/rouge/lexers/escape.rb]: eheh, yep, will remove.

**Comment by jneen** [comment]
I have rebased, and now I believe it's ready to merge pending yall's +1.

**Comment by jneen** [comment]
I also added escape functionality to the CLI, which is where I confess I actually need it at this moment, for use in the TeX formatter I'm working on.

**Pull request merged**